### PR TITLE
Add release workflow to automatically publish on PyPi

### PR DESCRIPTION
**Pull request by MaxDall**
This workflow adds the following jobs that trigger on any release with [type](https://docs.github.com/en/webhooks/webhook-events-and-payloads#release) `released`.

Jobs:

1. Test/Lint using existing workflows
2. Build distribution
3. Upload to TestPyPi
4. Test installation from TestPyPi
5. Upload to PyPi

### CODE REVIEW

**Comment by MaxDall** [comment]
I tested the action on a separate public test repository I created. You can find the corresponding run [here](https://github.com/MaxDall/test-commitizen/actions/runs/8436800068)

File: .github/workflows/publish-package.yml [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        uses: actions/setup-python@v5
+        with:
+          python-version-file: 'pyproject.toml'
+
+      - name: Install package
+        run: >- 
+          python3 -m 
+          pip install 
+          --index-url https://test.pypi.org/simple/ 
+          --extra-index-url https://pypi.org/simple/
```
Comment (dobbersc) [inline .github/workflows/publish-package.yml:92]: Is it intended to provide the extra index URL as the regular pypi distribution? I would have thought that the `test-distribution` job should only install fundus from testpypi and not from the regular pypi repository.

File: .github/workflows/publish-package.yml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        run: >- 
+          python3 -m 
+          pip install 
+          --index-url https://test.pypi.org/simple/ 
+          --extra-index-url https://pypi.org/simple/ 
+          fundus==${{ github.event.release.tag_name }}
+
+  publish-to-pypi:
+    name: Publish Python 🐍 distribution 📦 to PyPI
+    if: startsWith(github.ref, 'refs/tags/')
```
Comment (dobbersc) [inline .github/workflows/publish-package.yml:97]: Why is this check in the standard pypi publishing job but not in the one to testpypi? I thought it would not be necessary but maybe it is just as check to be sure?

File: .github/workflows/publish-package.yml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+      id-token: write
+
+    steps:
+      - name: Download all the dists
+        uses: actions/download-artifact@v3
+        with:
+          name: python-package-distributions
+          path: dist/
+      - name: Publish distribution 📦 to TestPyPI
+        uses: pypa/gh-action-pypi-publish@release/v1
```
Comment (dobbersc) [inline .github/workflows/publish-package.yml:70]: Just to be consistent with the other steps.

```suggestion
          path: dist/
          
      - name: Publish distribution 📦 to TestPyPI
        uses: pypa/gh-action-pypi-publish@release/v1
```

File: .github/workflows/publish-package.yml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+      id-token: write
+
+    steps:
+      - name: Download all the dists
+        uses: actions/download-artifact@v3
+        with:
+          name: python-package-distributions
+          path: dist/
+      - name: Publish distribution 📦 to PyPI
+        uses: pypa/gh-action-pypi-publish@release/v1
```
Comment (dobbersc) [inline .github/workflows/publish-package.yml:117]:
```suggestion
          path: dist/
          
      - name: Publish distribution 📦 to PyPI
        uses: pypa/gh-action-pypi-publish@release/v1
```

File: .github/workflows/publish-package.yml [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    name: Publish Python 🐍 distribution 📦 to TestPyPI
+    needs:
+      - build
+    runs-on: ubuntu-latest
+
+    environment:
+      name: testpypi
+      url: https://test.pypi.org/p/fundus
+    permissions:
+      id-token: write
```
Comment (dobbersc) [inline .github/workflows/publish-package.yml:61]:
```suggestion
      url: https://test.pypi.org/p/fundus
      
    permissions:
      id-token: write
```

File: .github/workflows/publish-package.yml [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    steps:
+      - name: Download all the dists
+        uses: actions/download-artifact@v3
+        with:
+          name: python-package-distributions
+          path: dist/
+      - name: Publish distribution 📦 to PyPI
+        uses: pypa/gh-action-pypi-publish@release/v1
+
```
Comment (dobbersc) [inline .github/workflows/publish-package.yml:118]: There are also a lot of deprecation notices on the [test run](https://github.com/MaxDall/test-commitizen/actions/runs/8436800068). But fixing them would be for another PR.

**Review by dobbersc — CHANGES_REQUESTED** [review]
This looks really useful! I have just some minor comments / questions.

File: .github/workflows/publish-package.yml [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        run: python3 -m build
+
+      - name: Store the distribution packages
+        uses: actions/upload-artifact@v3
+        with:
+          name: python-package-distributions
+          path: dist/
+
+  publish-to-testpypi:
+    name: Publish Python 🐍 distribution 📦 to TestPyPI
```
[the remark this answers is not in the document]
Comment (MaxDall) [reply .github/workflows/publish-package.yml:52]: Yes, that's basically it.

File: .github/workflows/publish-package.yml [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+          --index-url https://test.pypi.org/simple/ 
+          --extra-index-url https://pypi.org/simple/ 
+          fundus==${{ github.event.release.tag_name }}
+
+  publish-to-pypi:
+    name: Publish Python 🐍 distribution 📦 to PyPI
+    if: startsWith(github.ref, 'refs/tags/')
+
+    needs:
+    - test-distribution
```
[the remark this answers is not in the document]
Comment (MaxDall) [reply .github/workflows/publish-package.yml:100]: You're right!

File: .github/workflows/publish-package.yml (hunk 1 above)
Comment (MaxDall) [reply .github/workflows/publish-package.yml:92]: If we specify an `index-url` also site-packages (dependencies) are resolved using the same index, so we also need the `extra-index-url` to tell pip to look on `PyPi` for everything extra then the package itself.

File: .github/workflows/publish-package.yml [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+      - test
+      - lint
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v4
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version-file: 'pyproject.toml'
```
[the remark this answers is not in the document]
Comment (dobbersc) [reply .github/workflows/publish-package.yml:33]: Yea, I'm not sure there as well. I would just I the minimal version we support, e.g. 3.8.

File: .github/workflows/publish-package.yml (hunk 2 above)
Comment (MaxDall) [reply .github/workflows/publish-package.yml:97]: It isn't. Seems like I forgot to remove it.

File: .github/workflows/publish-package.yml (hunk 1 above)
Comment (dobbersc) [reply .github/workflows/publish-package.yml:92]: Oh, thanks makes sense. Thanks for the explanation.

File: .github/workflows/publish-package.yml (hunk 6 above)
Comment (MaxDall) [reply .github/workflows/publish-package.yml:119]: Yeah, I also noticed them but wasn't motivated enough to check if a simple version bump for all actions would solve the problem or if it's on the runners side.

File: .github/workflows/publish-package.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (MaxDall) [reply .github/workflows/publish-package.yml]: I pinned the version to `3.8`

**Review by dobbersc — APPROVED** [review]

File: .github/workflows/publish-package.yml (hunk 6 above)
Comment (dobbersc) [reply .github/workflows/publish-package.yml:119]: Lets just see how it is in future runs in the main repository and try a version bump later. Maybe open an issue for that?

**Review by dobbersc — APPROVED** [review]
